core/src/test/java/io/questdb/griffin/AlterTableAttachPartitionFromSoftLinkTest.java

```diff
@@ -0,0 +1,877 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.security.AllowAllCairoSecurityContext;
+import io.questdb.std.*;
+import io.questdb.std.datetime.microtime.TimestampFormatUtils;
+import io.questdb.tasks.ColumnPurgeTask;
+import io.questdb.test.tools.TestUtils;
+import org.junit.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+public class AlterTableAttachPartitionFromSoftLinkTest extends AbstractAlterTableAttachPartitionTest {
+
+    // some tests begin with: Assume.assumeTrue(Os.type != Os.WINDOWS);
+    // in WINDOWS the user performing the tests needs to have the 'Create Symbolic Links' privilege. 
+    // this privilege is not granted by default. in addition, if User Account Control (UAC) is on, and 
+    // the user has administrator privileges, tests must 'Run as administrator'.
+
+    private static final String expectedMaxTimestamp = "2022-10-18T23:59:59.000000Z";
+    private static final String expectedMinTimestamp = "2022-10-17T00:00:17.279900Z";
+    private static final String readOnlyPartitionName = "2022-10-17";
+    private static final long partitionTimestamp;
+
+    @Override
+    @Before
+    public void setUp() {
+        super.setUp();
+        Assert.assertEquals(TableUtils.ATTACHABLE_DIR_MARKER, configuration.getAttachPartitionSuffix());
+        Assert.assertFalse(configuration.attachPartitionCopy());
+    }
+
+    @Test
+    public void testAddColumn() throws Exception {
+        assertMemoryLeak(FilesFacadeImpl.INSTANCE, () -> {
+            final String tableName = testName.getMethodName();
+            createTableWithReadOnlyPartition(tableName, () -> {
+                        try {
+                            executeOperation("ALTER TABLE " + tableName + " ADD COLUMN ss SYMBOL", CompiledQuery.ALTER);
+                            assertInsertFailsBecausePartitionIsReadOnly(
```

Comment:
I don't see tests which insert many rows in read-only and normal partitions in 1 transaction, please add

Reply:
Also please add tests where the last partition is read-only and append of the rows fails into the last partition (no O3 involved).

Reply:
I have created a test that creates 5 partitions, makes the first four read only then drops the last partition. This makes the previously read-only partition now active, and thus you can write to it. It is an edge case. I would expect that you can always append to a table (even if it is in a location pointed to by a soft link). But I am not sure if this thinking is correct, what do you think @ideoma ?

(working on the transaction test)